File: skills/code-architect/references/code-examples/csharp/Singleton.cs
// Singleton — ensure one instance with a global access point.
using System;

namespace Patterns.Singleton
{
    sealed class Registry
    {
        // Thread-safe lazy initialization via Lazy<T>.
        private static readonly Lazy<Registry> _instance = new(() => new Registry());
        public static Registry Instance => _instance.Value;

        private int _counter;

        private Registry() { } // prevent external construction

        public int Next() => ++_counter;
    }

    static class Demo
    {
        public static void Run()
        {
            Registry a = Registry.Instance;
            Registry b = Registry.Instance;
            Console.WriteLine($"Same instance: {ReferenceEquals(a, b)}");
            Console.WriteLine(a.Next());
            Console.WriteLine(b.Next());
            Console.WriteLine(Registry.Instance.Next());
        }

        public static void Main() => Run();
    }
}

File: skills/code-architect/references/code-examples/java/Singleton.java
// Singleton — ensure one instance with a global access point.

public class Singleton {

    static final class Registry {
        private static volatile Registry instance;
        private int counter = 0;

        private Registry() {}

        // Thread-safe lazy initialization (double-checked locking).
        static Registry get() {
            if (instance == null) {
                synchronized (Registry.class) {
                    if (instance == null) instance = new Registry();
                }
            }
            return instance;
        }

        int next() { return ++counter; }
    }

    public static void main(String[] args) {
        Registry a = Registry.get();
        Registry b = Registry.get();
        System.out.println("same instance: " + (a == b));
        System.out.println("ticket " + a.next());
        System.out.println("ticket " + b.next());
        System.out.println("ticket " + Registry.get().next());
    }
}
